### ISSUE TITLE
Capture + Display all logs for everything that happens in Cypress

### ISSUE THREAD

**Issue by brian-mann** [cypress-io/cypress#448]
### Problem
Many of our users express the desire to see `console.log` statements when running headlessly. Typically you'd be able to see these when you're working in Dev Tools (in headed GUI mode).

We can pretty easily display `console.log` statements by mixing them `stdout` when running headlessly, but we don't think this is really a "true" solution. It's kind of a hack and it still doesn't really tell you "everything" that's happening. For instance, when a `console.log` happens, how would you know when it happened in relation to everything else that's going on in a test?

So, we think to **truly** solve this problem, we need a comprehensive way of displaying not just `console.log` statements but also capturing:

- Network Traffic
- Commands
- Retries
- Under the hood Cypress wizardry
- DOM snapshots on failure

By providing **all** of this, the entire picture of what happened during a test is revealed and now you actually have even **more** information at your disposal than even what the GUI headed version provides you.

### Solution
We're going to make these logs available both in our Dashboard, and in the Desktop App itself.

Once we capture and parse these logs, it will unlock an enormous amount of power in Cypress. We'll also be able to analyze the logs and deliver things like:

- Analytics into your tests
- Comparing a failing test to a previously passing one
- Diffing the logs for changes in response bodies, etc
- Understanding why tests are slow
- Discovering command anti patterns (like explicit waits in your code)

The logs should also be interactive - for instance they should "sync" up to the video and enable you to playback everything that actually happened in a Cypress test.

Here is an example comp of what we're going for.

<img width="1349" alt="logs 1" src="https://cloud.githubusercontent.com/assets/1268976/23758542/2479a79c-04b8-11e7-87c7-7d3313bf5161.png">

We're still working on `start` and `end`, and also the communication that a log event acts as "a stream" but that events are connected to other events.

For instance when a HTTP request starts - that's an event. When the response comes back, thats a separate event. However these are connected and you likely need to be able to view the data for both events from either the `start` or `end` event.

Another example:

A command starts. It retries 65x times. It eventually finishes. The duration of the retry events, the reason its retrying, and the eventual delta between a start and and end needs to be communicated intelligently.

These log files end up being a massive amount of data, but we already have the infrastructure in place to capture, parse and deliver them. At this point we're just iterating on a UI that's intuitive and doesn't overwhelm the user.

We're going to be providing a new `Logs` tab, but also for individual test failures, we'll just be providing the logs for that one independent test.

**Comment by brian-mann** [comment]
@bahmutov what do you think?

**Comment by bahmutov** [comment]
Perfecto! 😍

**Comment by andyshora-qb** [comment]
@brian-mann excellent. Really looking forward to getting this. We're spending lots of time trying to debug why our compiled JS is not working when served static via Cypress. We cannot replicate anywhere else, so one peek at the logs would like solve everything!

**Comment by danielnass** [comment]
Hi guys!

Amazing feature! Any news about that?

Thanks

**Comment by cristopher-rodrigues** [comment]
**AMAZING**!

**Comment by lukemadera** [comment]
@brian-mann I've been banging my head against a wall trying to upgrade to cypress 0.20.1. It's been days - things work fine in the GUI but fail from the console. And not being able to log anything is infuriating. After super old school commenting out code one line at a time until it passes again to isolate the line, I've found that it is `cy.request` that fails. But I don't know how to fix it yet, and I can't log anything so have no idea what the issue is. Adding all the features you noted in this ticket sounds fine, but we need a simple way to debug locally, in the command line itself. How can we get console.log (or cy.log) to show in stdout when running headlessly? We're blocked from using Cypress right now until we fix that because our CI is failing now due to this other error, which is supposedly fixed in 0.20 https://github.com/cypress-io/cypress/issues/517

**Comment by bahmutov** [comment]
@lukemadera have you tried using `DEBUG` variable to see messages from Cypress? https://docs.cypress.io/guides/guides/debugging.html#Debug-the-Command-Line

**Comment by brian-mann** [comment]
1. Is it failing locally or only in CI?
2. If failing locally be sure to run in `--headed` mode so you can see what's going on.
3. You can also switch to using `--browser chrome` from the CLI (in local and in CI) if it fails only in electron. Electron is technically a different browser from your normal Chrome.
4. As per getting events out of the browser - that is not necessarily a straightforward thing. What exactly would you like us to send to `stdout`? Internal information about what Cypress is doing? You will see hundreds if not thousands of events being fired over the course of a run. And that's just what **Cypress** is doing. Nothing from your application like network requests, console output, etc would be sent unless we chose to do that.
5. I would like to run an experiment here. Open up Cypress in GUI mode and then open the console and follow the instructions here: https://docs.cypress.io/api/events/catalog-of-events.html#Logging-All-Events.

That will send every internal event that Cypress fires to the console. Look at this and tell me if it is helpful. If the answer is yes, we could pretty easily get that out of the browser and into `stdout`. If it is not helpful then we'd need to figure out what is. Browsers are enormously complex and trying to distill what's happening down into text is not that simple. What exactly do you want to see? What would be helpful?

**Comment by brian-mann** [comment]
Also one last important piece of information. Why aren't you just recording your runs? Even though its headless you can see exactly what is going on since we capture a movie and screenshots. What about that information is not enough to see what's going on and be able to reproduce locally?

**Comment by brian-mann** [comment]
Or at the very least why aren't you watching the video from headless runs or looking at screenshots?

**Comment by lukemadera** [comment]
@bahmutov Yes I've tried debug. That gives general info, not info I generate and want to see.

@brian-mann thanks for the speedy response. There are 2 failures / issues right now.
A. on CI, v0.19.x no longer runs (headless)
B. locally, v0.20.x runs in GUI but fails in headless.

2. All the tests pass in the GUI so unless I'm missing something, the error won't happen in headed mode, so I can only debug in headless mode.

3. Again, it's a general request issue (I think - still need logging), not a browser issue.

4. I can reproduce it locally 100% of the time, in headless mode. Works 100% of the time in GUI mode. These are tests that worked fine for months, and they work in v0.19. Just not when upgrading to v0.20.x

5. I'm simply looking to be able to do `console.log()` and have it show up on the command line when I do `./node_modules/.bin/cypress run --spec cypress/integration/locations/locations-list.js`

6. while all that data MAY be helpful, right now I just need ONE (or a few) pieces of data, that I choose from `console.log()`

7. I'm not even getting to the screenshots or videos really, and I've already figured out the problem (it's in our data setup phase, trying to login via cy.request). It's not an actual test that is failing, we're not even making it to the tests. So I'm not sure how videos / screenshots would help. It's literally the first step that's failing, in setup, before tests even start.

If this is still unclear I can record a video of the test in the GUI and the test in the command line - one finishes and passes, one just gets stuck and never does anything (or times out).

**Comment by brian-mann** [comment]
Okay let me clear up a few things.

When we say GUI mode we're talking about launching cypress from `cypress open`. In that mode the tests never "complete" and you don't get anything on `stdout`.

When you run from the CLI via `cypress run` then the tests complete, you get `stdout` and an exit code.

Cypress itself **does** work differently in these two modes. When running from the CLI it makes internal optimizations to be able to run all the tests - ie it will run faster than the GUI mode. The GUI mode is focused on debuggability and interaction.

My point is this: using `--headed` from the CLI simply shows you the browser. It will behave identically to running headless. One shows the Electron browser window, and the other doesn't. That's it.

In fact you could just launch Chrome via the CLI with `--browser chrome` and then you'll be using the same two browsers in GUI mode and in CLI mode with the only difference that one runs to completion, the other doesn't.

The screenshots and video would still help in this situation, albeit I understand the need for `console.log`. At the very least they would show you precisely which step and command is failing without you having to guess in the dark.

**Comment by brian-mann** [comment]
Also at this very moment you could do this trick here: https://github.com/cypress-io/cypress/issues/300#issuecomment-321587149

That will get all console.log's to show up in the command log, which will stringify and print their arguments. You could then watch the video / screenshot to see what the information is.

That's what I would do today when running in CI. Of course locally just show the browser and iterate on it there.

**Comment by brian-mann** [comment]
I opened a couple new issues to do what you're asking with exporting the `console` output, and some improvements to the debugging the driver.

**Comment by lukemadera** [comment]
YES! The `--headed` works and I can see console.log output, thanks so much @brian-mann ! It does still close automatically at the end, which means I can't see the logs persisted, so I have to add a wait at the end. Any way to prevent the browser from auto-closing?

However, I'm still stuck at the same `cy.request` line. It just doesn't ever complete. In digging into it, `cy.request` works fine if it's called BEFORE `cy.readFile`. If it's used after (within the .then()), it just never returns.
In trying to reproduce it, if I simplify things it seems to work, so I'm not clear on what the issue is.
If I just do a request (using the npm `request` module instead of cy.request, I get a CORS error).`

Luckily, the request I'm attempting is to logout so in this case I can just skip that request all together, and then things work fine. Now tests appear to be passing again.. Locally at least. Going to try in CI now.

Not sure if the issue is "fixed" but it's working for me now it seems. Thanks for all the speedy responses and help!

**Comment by brian-mann** [comment]
If you want to iterate without it close, just open up Electron in the GUI. Just select that browser instead of Chrome. You may see the same issue.

As I mentioned if that's the case just switch to booting chrome from the CLI `--browser chrome`

**Comment by rovansteen** [comment]
This would be a really nice addition. Right now my tests are failing on the CI and it looks like it has something to do with cookies but because Cypress is basically a black box right now on CI I have no way of figuring out what's going on. Hope this feature will arrive soon, keep up the good work 👏🏼

**Comment by brian-mann** [comment]
Related to #699 and #700.

Those will land pretty soon and should help alleviate these problems.

**Comment by brandonb927** [comment]
Just want to throw my 2 cents in, I'm also having an issue with (what I can only imagine is cookie-related) the CI build vs local running that @rovansteen is having, and not having a way to see what network requests are made is making it challenging to debug.

So far I'm loving Cypress though, you folks are doing amazing work and I'm more than willing to fork over our company credit card for a paid plan in the future!

**Comment by RRMoelker** [comment]
Hi, loving Cypress. But I am curious. Are these issues still being addressed?

I see a lot of comments that these issues are being worked on, like: "x and y. Those will land pretty soon".  But this seems to have been in development for nearly a year now.

Respectfully, if we can't debug our CI than this isn't really a viable product for the long term. And we would rather know sooner than later since E2E tests are very costly.

**Comment by aboutlo** [comment]
@brian-mann do you have any update on this one? In October you said soon. I know your backlog is very long but CI logs are quite important. Every time there is an issue on the CI we need to run the tests again locally in order to get the logs. 
Thank you for your understanding 🙏

**Comment by szabyg** [comment]
I'm wondering if anything is happening here, it's way too quite for a feature all projects need at the point where they have to maintain tests and the CI console output and the test results can't tell what went wrong except the last exception message.

Sounds like a hack but if you 
1. start an HTTP server in the background with a log endpoint (the endpoints simply writes everything on its console)
2. patch the cy.log() command to make a request to your endpoint with the arguments it got

You should at least see the cy.log messages. It would be much more elegant if it goes through the existing cypress internal websocket though.

Anyone here tried something like this before?

**Comment by bahmutov** [comment]
For everyone who wants to get the `cy.log` and other Cypress commands this might be helpful https://github.com/bahmutov/cypress-failed-log

**Comment by jennifer-shehane** [comment]
Since 3.0, we're collecting more data related to the timings, body, and hooks run for recorded tests. 

This issue has been put on hold as there was a great deal of work to be done in the server to gather all of the data necessary. It is still on our roadmap for the future - and there has been *some* work done on it.

**Comment by souljacker** [comment]
+1 for this being implemented

**Comment by zhex900** [comment]
Hi,
Is the dashboard logging tab still in the development plan?

**Comment by zachfeldman** [comment]
We just had this issue, where tests pass locally, but fail on CI. We have no way of knowing why they fail as we just get a White Screen of Death and no console.log output. I'm at a loss for any way to know what error is being thrown as I can't even reproduce it locally. Any other tips for jury-rigging console logging on Travis?

**Comment by amirrustam** [comment]
This feature is part of our [Dashboard Service Roadmap](https://docs.cypress.io/guides/references/roadmap.html#Dashboard-Service), and it is **now under active development**.

**Comment by kamituel** [comment]
@amirrustam Does it mean it will be available only when using Dashboard (and not when using standalone `cypress run` without test recording)?

**Comment by amirrustam** [comment]
@kamituel we've considered both use cases. No additional details available at this time. So stay tuned. Generally speaking, using the Dashboard enables ancillary benefits that are geared towards pin-pointing test failure root causes and faster debugging.

**Comment by Vandivier** [comment]
Cypress newcomer and senior developer here. My opinion fwiw: The dashboard sounds like a great feature but I don't think it substitutes for intuitive logging.

What's intuitive for a developer? Using devtools console and/or printing to command line, just like the other testing frameworks do. Given the cool left-hand side bar I'd also expect logs to turn up there. In one example, I ran `cy.log` just before a failed assertion and unexpectedly got no info.

As for intuition, I'd also think `cy.log(object)` working much like `console.log(object)` would be neato.

Cheers to the work done so far!

**Comment by Pomax** [comment]
The dashboard is not available during automated testing runs, such as running Cypress on Travis-CI, where it is _critically important_ that there is a way to be able to log output to `stdout`.

Can we still get something like `cy.stdout.write`?

**Comment by jennifer-shehane** [comment]
Hey @Pomax 

>The dashboard is not available during automated testing runs, such as running Cypress on Travis-CI

Could you explain a bit more by what you mean in the statement above? Is your company not able to record to the Dashboard? I know there have been requests for [On Prem](https://github.com/cypress-io/cypress/issues/3095) due to security restrictions, for example. 

Having the ability to access the raw logs is certainly something our team is considering.

**Comment by Pomax** [comment]
We're using Cypress through Travis: it runs a fully contained test pass during CI and the only thing we get to see is the log that travis-ci gives you. The only thing Cypress gets to do is generate its DOM snapshot, and that data is then sent off to a second, independent CI service that runs a million browsers at a million resolutions, which can all be checked in parallel. The only output we get is stdout log on Travs, so not being able to put in `cy.console.log('waiting for React to finish..')` makes it incredibly hard to determine where things might get stuck on CI (which by definition has a different run profile than local dev machines).

**Comment by mitar** [comment]
OK, while we wait for this perfect solution described in OP to arrive, I would really like to vote for a very simple solution of redirecting all browser console to the terminal during CI run. It is really hard to debug anything currently when it fails in CI. I would love to simple see in terminal all logging calls I might have inside tests themselves and also from my app. I mean, we already have those logging calls for debugging with a browser, let's just log them to the terminal as well.

**Comment by jennifer-shehane** [comment]
@bahmutov created a plugin that will print Cypress commands to the terminal - not exactly the browser console logs wanted, but maybe will help someone. https://github.com/bahmutov/cypress-failed-log

We've also been expanding on our Debugging doc - adding some weird tricks in there that could help some people. https://on.cypress.io/debugging

**Comment by aboutlo** [comment]
First apologies for the comment which is not trying to fix the actual problem. 🙏 
Then let's say cypress is really cool and I really liked to use it 🙌 

However, I can't suggest / use it to anyone until this is fix it. 
The workaround is fine but far from being a good solution.
Being by default blind in CI is a big deal.

**Comment by MaaikeFox** [comment]
What I now do is keep cy.logs really short so that I can hopefully read them from the screenshots that are available on the Dashboard. But it is a silly solution of course. It would be so helpful if the cy.logs would be available on the Dashboard

**Comment by saas2813** [comment]
I want to assert that the Console log is empty, or only contains previously approved lines in the CI build. The optimal implementation would give a simple textfile with timestamped direct copies from the console output. 
These logs could be prepended by the testcase name and current URL for context, adding more stuff would be noice.

As others have commented, it feels rather bad that the console log and Cypress command log can not be saved to files (separately or together, by configuration) 2.5 years after the issue (#300) came up.

**Comment by rodya-mirov** [comment]
It's been 2.5 years since this was brought up and the fix everyone wants was declared "easy." All of the suggested fixes around using the dashboard are problematic for reasons described thoroughly in this issue. Is there any movement on this?

**Comment by bahmutov** [comment]
Current workarounds: use these 2 plugins
- https://github.com/flotwig/cypress-log-to-output
- https://github.com/bahmutov/cypress-failed-log

**Comment by alfaproject** [comment]
Just stumbled upon this need ):

**Comment by waitsangbv** [comment]
https://github.com/flotwig/cypress-log-to-output workaround only works on Chrome. But chrome headless does not work and does not support video recording... its going round and round the loop for a very simple request (i.e. proper console,log)

**Comment by hvuong-sigma** [comment]
@jennifer-shehane 

I don't see the feature listed in the Cypress Roadmap anymore and the two suggested workarounds aren't ideal, at least for me. Is there still planned work for this? We're fully integrated with Cypress and Cypress Dashboard but we're having trouble debugging issues without having the console.log available.

**Comment by RamiaSaidawi** [comment]
any update on how we can get console.log showing when we use "cypress run" ?
e are thinking of moving to cypress and this can be a show stopper :(

**Comment by x-yuri** [comment]
In case you're using Electron browser you can see your `console.log()` statements with `DEBUG=cypress:electron npx cypress run`. More on it [here][1]. For Chrome the abovementioned [`cypress-log-to-output`][2] plugin should help

[1]: https://github.com/cypress-io/cypress/issues/700#issuecomment-562840560
[2]: https://github.com/flotwig/cypress-log-to-output

**Comment by craig-dae** [comment]
Any updates on this? I have a cicd test that fails, that I can't reproduce locally, and I'm burning through my Github Action minutes.

**Comment by jennifer-shehane** [comment]
This issue is still in the 'proposal' stage, which means no work has been done on this issue as of today, so we do not have an estimate on when this will be delivered.

## Current workarounds: use these plugins

- https://github.com/flotwig/cypress-log-to-output - Prints all browser console logs to the terminal while running Cypress tests. Currently, only Chrome is supported.
- https://github.com/bahmutov/cypress-failed-log - Saves the Cypress test command log as a JSON file if a test fails.
- https://github.com/archfz/cypress-terminal-report - Logs cypress commands, route request data and browser console errors and warnings to terminal when tests fail on CI.

**Comment by AbrahamBrookes** [comment]
I'd like to write my own little helper for this, what command do I use to print arbitrary strings to the CLI when using `cypress run`? Or is that the underlying issue here - that there is none?

**Comment by braveheart55** [comment]
Jumping in on the bandwagon calling for better logging especially on failure when matching json objects.  I'm looking for a test framework for REST API  testing where the lack of logging/reporting prevents me from choosing cypress

**Comment by csvan** [comment]
> Jumping in on the bandwagon calling for better logging especially on failure when matching json objects. I'm looking for a test framework for REST API testing where the lack of logging/reporting prevents me from choosing cypress

That's not really the usecase Cypress seeks to meet. If you want to test APIs, you should probably be looking at Postman.

https://www.postman.com/api-platform/api-testing/

**Comment by braveheart55** [comment]
Thank you for the link.  I am a long time postman user, however I didn't know about their automated test capability.  Thank you

**Comment by rennerg** [comment]
@brian-mann any updates on this?

**Comment by lgschell** [comment]
Hi @brian-mann could you make any comment on this ticket ? Would be great to have some news on its progress since it was raised a long time ago (March 2017). Thank you very much.

**Comment by hadrien-toma** [comment]
I really like Cypress because of the attention put on developer experience so I will share my feeling hopping it will motivate / bounce this issue:
As a DevOps developer I can tell that this feature will considerably enhance my experience, for the moment I need to use "cy.writeFile" to collect all logs and then I need to download artifacts to see this written file, which is relatively painful...

**Comment by dembeleu** [comment]
This is so frustrating .... anything new on this one?

**Comment by melibe23** [comment]
Do we have news on this? It would be awesome to add it.

**Comment by aomaryus** [comment]
When is this feature coming. We really need it.
Thank you.

**Comment by willischu** [comment]
➕

**Comment by hannaace** [comment]
➕ 1

**Comment by tchiaspko** [comment]
+1

**Comment by PJEstrada** [comment]
+9000

**Comment by aomaryus** [comment]
I need this feature. Been waiting on it.

**Comment by anastasiyamazheika** [comment]
+1

**Comment by josh803316** [comment]
➕  1

**Comment by Phonesis** [comment]
I am amazed this is still not a feature of Cypress tbh. Is there really no way to show cy.log() entries in the official results dashboard?

**Comment by filiphric** [comment]
@Phonesis This may not be the exact thing you are looking for, but the way I display `cy.log()` entries in dashboard is by appending them to error message, you can find the code snippet at the end of [this blogpost](https://filiphric.com/improve-your-error-screenshots-in-cypress)

**Comment by ChelseaChilders** [comment]
It's now been over 5 years and this is still not a feature. Being able to view the console.logs (and such) would extremely help in debugging failing tests.

**Comment by csvan** [comment]
There are tons of workarounds in this thread alone, have you considered any of them?

**Comment by richardszanyi-kasa** [comment]
That would be awesome, also if we could group commands such as cy. session, that would be beneficial as well. :)

**Comment by usmanmustafa33** [comment]
➕

**Comment by lmiller1990** [comment]
A lot of work happened over the last little while to enable us to finally work on something like this. It's not a lot more practical to implement something like this due to various under-the-hood improvements to the App and Cloud platforms over the last year or two. I hope we can share some updates and progress soon.

In the meantime, some debugging-in-CI techniques I've found useful:

> Is there really no way to show cy.log() entries in the official results dashboard?

The way I've done this historically is by using `cy.log` and `cy.screenshot`. You can do `cy.screenshot(..., { capture: 'runner' })` to include the command log in the screenshot. 

You can pass `JSON.stringify(<value>, null, 4)` and get some pretty elaborate logging going. Not really a fully featured solution, but it's helped me out a ton when debugging.

Screenshot docs: https://docs.cypress.io/api/commands/screenshot#Arguments

**Comment by Opalexiss** [comment]
I am looking for this as well, and none of the workarounds are usable for me.
My app outputs various information to the console while in debugging mode, which is enabled for Cypress.
In a test, I would like to do something simple such as click on a button, then make an assertion based on what was immediately logged to the developer console afterwards.

The only actual workaround for this is to stick a hidden div somewhere in the DOM that mirrors everything that is being output to the console. My apps communicate internal states and variables for tests, and these are not always accessible from the DOM itself. I would like to assert based on these console.log messages.

**Comment by lmiller1990** [comment]
You can spy/stub the console:

- https://stackoverflow.com/questions/53898085/check-if-an-error-has-been-written-to-the-console
- https://www.youtube.com/watch?v=zRAjO6rccfI

**Comment by M4RcB4** [comment]
> This issue is still in the 'proposal' stage, which means no work has been done on this issue as of today, so we do not have an estimate on when this will be delivered.
> 
> ## Current workarounds: use these plugins
> * https://github.com/flotwig/cypress-log-to-output - Prints all browser console logs to the terminal while running Cypress tests. Currently, only Chrome is supported.
> * https://github.com/bahmutov/cypress-failed-log - Saves the Cypress test command log as a JSON file if a test fails.
> * https://github.com/archfz/cypress-terminal-report - Logs cypress commands, route request data and browser console errors and warnings to terminal when tests fail on CI.

@brian-mann @bahmutov @lmiller1990 Any official work being done on this feature? I will try the plugins mentioned here. I would also love to keep in step with future improvement. I think this is an immensely valuable feature.

**Comment by lmiller1990** [comment]
@M4RcB4 Yes, there is. I don't have an exact date, but it's an active and high priority project. Please stay tuned - more info will come as soon as it is ready for testing!

The goal is to capture *all* the things - console, network etc. As a long time user, I am quite excited - debugging with videos is good, but this will be better.

**Comment by boweiliu** [comment]
Hey, is the intention for the logs to be only available via cypress dashboard, or would they also be printed to the terminal/saved to a local .txt as well?

**Comment by ryanthemanuel** [comment]
The project that @lmiller1990 mentioned will be only available in the cypress dashboard.

**Comment by Opalexiss** [comment]
So, still no way to make assertions against console output.

**Comment by lmiller1990** [comment]
@Opalexiss you can definitely make assertions against the `console` methods: https://glebbahmutov.com/cypress-examples/recipes/check-console-logs.html

**Comment by agoldis** [comment]
Here's a free, open-source plugin that you can use to capture everything, including DOM changes (real HTML), browser console messages and network requests. You can then replay everything in a web-based player. Think of Playwright Traces, but for Cypress.

https://github.com/currents-dev/cypress-debugger

**Comment by chrisbreiding** [comment]
Closed by https://github.com/cypress-io/cypress/pull/27040

**Comment by karlhorky** [comment]
@chrisbreiding nice, thanks! Although, the PR description on that #27040 PR is a bit unclear - which PR in v13.0.0 actually addresses this issue #448? 

I couldn't figure that out from either the PR description or [the v13.0.0 changelog](https://github.com/cypress-io/cypress/blob/d75dab107e3903da47b94794b40c55e8bf6e7ab7/cli/CHANGELOG.md)

**Comment by chrisbreiding** [comment]
@karlhorky I can understand how it might be a bit confusing. #27040 was a PR for a long-running branch that collected a number of PRs for the v13 release. That includes several PRs that address this issue. For the most part, however, those changes are not user-facing, so they're not called out in the v13 PR or the changelog. Most of the work for this is under-the-hood (capturing all the things!).

While the changelog published with this repo doesn't mention it, the [changelog](https://github.com/cypress-io/cypress-documentation/pull/5315/files#diff-d135fdaff151adc69e9453a47526a87201a262b56134f936dc9ead92e4622d59R9-R16) that will be published to the docs highlights the functionality this will enable for Cypress Cloud.

**Comment by jennifer-shehane** [comment]
Released in [Cypress 13.0.0](https://docs.cypress.io/guides/references/changelog#13-0-0) as Test Replay. Please read our [Blog](https://www.cypress.io/blog/2023/08/29/announcing-cypress-13-test-replay/) for more details.

**Comment by karlhorky** [comment]
Nice, thanks for the additional background, looks exciting!

**Comment by tchaton** [comment]
@jennifer-shehane I am not sure Test Replay was the actual requested feature. 

Personally, I want to be able to dump console.log, network tabs to a file, so we I can read it to understand what went wrong. 

Best,
T.C

**Comment by csvan** [comment]
Kudos to the Cypress team for Test Replay which is an awesome feature, but like @tchaton states this does NOT address this issue, at all.

**Comment by comp615** [comment]
@jennifer-shehane what's the story for replay around non-cloud CI/CD setups? 

I'm a bit worried this is a dark pattern change designed to remove useful functionality from that set of users, especially when viewed as coupled with the removal of `videoUploadOnPasses`, for instance. Happy to take the tin foil hat off, but I didn't see any discussion here or on other threads of allowing an "asset" to be captured in CI that would let people view the runs locally.

**Comment by lmiller1990** [comment]
@comp615 Was something **removed**? I don't work at Cypress, but I still use it a lot -- my understanding is the Test Replay is for Cloud users. You can still record videos to your local machine, review them after a failure, as part of the OSS offering.

**Comment by viqcinese-axur** [comment]
Test Replay is not a replacement for adequate error logs. I still can't get over the fact that Cypress is a testing tool that runs in a CLI but does not inform the user what went wrong when a test fails.
